flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/TableUtil.java

```diff
@@ -0,0 +1,429 @@
+	 * @param selectedCols the selectedCols to assert.
+	 */
+	public static void assertSelectedColExist(String[] tableCols, String... selectedCols) {
+		if (null != selectedCols) {
+			for (String selectedCol : selectedCols) {
+				if (null != selectedCol) {
+					if (-1 == findColIndex(tableCols, selectedCol)) {
+						throw new IllegalArgumentException(" col is not exist " + selectedCol);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Check whether colTypes of the <code>selectedCols</code> is numerical, if not, throw exception.
+	 *
+	 * @param tableSchema  TableSchema
+	 * @param selectedCols the selectedCols to assert.
+	 */
+	public static void assertNumericalCols(TableSchema tableSchema, String... selectedCols) {
+		if (selectedCols != null && selectedCols.length != 0) {
+			for (String selectedCol : selectedCols) {
+				if (null != selectedCol) {
+					if (!isNumber(findColType(tableSchema, selectedCol))) {
+						throw new IllegalArgumentException("col type must be number " + selectedCol);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Check whether colTypes of the <code>selectedCols</code> is string, if not, throw exception.
+	 *
+	 * @param tableSchema  TableSchema
+	 * @param selectedCols the selectedCol to assert.
+	 */
+	public static void assertStringCols(TableSchema tableSchema, String... selectedCols) {
+		if (selectedCols != null && selectedCols.length != 0) {
+			for (String selectedCol : selectedCols) {
+				if (null != selectedCol) {
+					if (!isString(findColType(tableSchema, selectedCol))) {
+						throw new IllegalArgumentException("col type must be string " + selectedCol);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Check whether colTypes of the <code>selectedCols</code> is vector, if not, throw exception.
+	 *
+	 * @param tableSchema  TableSchema
+	 * @param selectedCols the selectedCol to assert.
+	 * @see #isVector(TypeInformation)
+	 */
+	public static void assertVectorCols(TableSchema tableSchema, String... selectedCols) {
+		if (selectedCols != null && selectedCols.length != 0) {
+			for (String selectedCol : selectedCols) {
+				if (null != selectedCol) {
+					if (!isVector(findColType(tableSchema, selectedCol))) {
+						throw new IllegalArgumentException("col type must be string " + selectedCol);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Return the columns in the table whose types are string.
+	 *
+	 * @param tableSchema TableSchema
+	 * @return String columns.
+	 */
+	public static String[] getStringCols(TableSchema tableSchema) {
+		return getStringCols(tableSchema, null);
+	}
+
+	/**
+	 * Return the columns in the table whose types are string and are not included in the excludeCols.
+	 *
+	 * <p>If <code>excludeCols</code> is null, return all the string columns.
+	 *
+	 * @param tableSchema TableSchema.
+	 * @param excludeCols The columns who are not considered.
+	 * @return string columns.
+	 */
+	public static String[] getStringCols(TableSchema tableSchema, String[] excludeCols) {
```

Comment:
question on this and all below `getXXXCols` API. why is the fields being get needs to have an "excludeCols" instead of include?

Reply:
Take the "getNumericCols" API as an example: when there are a number of feature columns and we want to use all the numerical columns as feature columns except the label column, this API is used.